packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1159 @@
+///
+/// ## Application architectural design
+///
+/// An application can have zero, one, or many [Router] widgets, depending on
+/// its needs.
+///
+/// An application might have no [Router] widgets if it has only one "screen",
+/// or if the facilities provided by [Navigator] are sufficient.
+///
+/// A particularly elaborate application might have multiple [Router] widgets,
+/// in a tree configuration, with the first handling the entire route parsing
+/// and making the result available for routers in the subtree. The routers in
+/// the subtree do not participate in route information parsing but merely take the
+/// result from the first router to build their sub routes.
+///
+/// Most applications only need a single [Router].
+///
+/// ## URL updates for web applications
+///
+/// In the web platform, it is important to keeps the URL up to date with the
+/// app state. This ensures the browser constructs its history entry
+/// correctly so that its forward and backward buttons continue to work.
+///
+/// If a app state change leads to [Router] rebuilds (usually triggered by a
+/// navigating method), the [Router] will detect such event and intend to
+/// report the new route information back to the web engine by using the
+/// [RouterDelegate.currentConfiguration] and the [RouteInformationParser.restore].
+/// Providing implementations of these two methods in the subclasses enables the
+/// [Router] widget to successfully update the URL in the web engine.
+///
+/// You can also choose to opt out of URL updates. Simply ignore the
+/// [RouterDelegate.currentConfiguration] and the [RouteInformationParser.restore]
+/// without providing the implementations will prevent the [Router] from
+/// reporting the URL back to the web engine. You will want to opt out in these
+/// cases:
+///
+/// If you are not writing a web application.
+///
+/// If you have multiple router widgets in your app, then only one router widget
+/// should update the URL (Usually the top-most one).
+///
+/// If your app does not care about the in-app navigation using the browser's
+/// forward and backward buttons.
+///
+/// Otherwise, we strongly recommend implementing the
+/// [RouterDelegate.currentConfiguration] and the [RouteInformationParser.restore]
+/// to provide optimal experience in the web application.
+class Router<T> extends StatefulWidget {
+  /// Creates a router.
+  ///
+  /// The [routeInformationProvider] and [routeInformationParser] can be null if this
+  /// router does not depend on route informations. A common example is a sub router
+  /// that builds its content completely relies on the app state.
+  ///
+  /// If the [routeInformationProvider] is not null, the [routeInformationParser] must
+  /// also not be null.
+  ///
+  /// The [routerDelegate] must not be null.
+  const Router({
+    Key key,
+    this.routeInformationProvider,
+    this.routeInformationParser,
+    this.backButtonDispatcher,
+    @required this.routerDelegate,
```

Comment:
nit: per the style guide, the order of the constructor arguments and the order of the properties should be the same.  We should make that order make sense when reading the code (e.g. routeInformationProvider and routeInformationParser should always be together).